src/Analyser/MutatingScope.php

```diff
@@ -1621,12 +1621,18 @@ private function resolveType(string $exprString, Expr $node): Type
 			if ($this->hasExpressionType($node)->yes()) {
 				return $this->expressionTypes[$exprString]->getType();
 			}
-			return $this->initializerExprTypeResolver->getClassConstFetchTypeByReflection(
+			$constantType = $this->initializerExprTypeResolver->getClassConstFetchTypeByReflection(
 				$node->class,
 				$node->name->name,
 				$this->isInClass() ? $this->getClassReflection() : null,
 				fn (Expr $expr): Type => $this->getType($expr),
 			);
+
+			if ($this->isInTrait() && !$constantType->isClassStringType()->yes()) {
+				return $constantType->generalize(GeneralizePrecision::lessSpecific());
+			}
+
+			return $constantType;
```

Comment:
No idea what this code is going to do, can you verify that with NodeScopeResolverTest?

Reply:
the idea is, when analyzing in a trait we generalize constant values to their base type, e.g. `'hallo'` -> `string`.

that way we don't run into the always true/false errors